In [1]:
import re
import os
import glob
import zipfile
import requests
from urllib.request import urlretrieve
import json
import pandas as pd
from memory_profiler import memory_usage
import numpy as np

In [2]:
%load_ext rpy2.ipython
%load_ext memory_profiler

# Changing `dtype` of the data

In [3]:
# Loading the combined dataframe to the memory
df = pd.read_csv("../data/processed/combined_data.csv")
df.head()

time    lat_min    lat_max   lon_min   lon_max  \
0  1889-01-01 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
1  1889-01-02 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
2  1889-01-03 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
3  1889-01-04 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
4  1889-01-05 12:00:00 -35.439867 -33.574619  141.5625  143.4375   

   rain (mm/day)            model  
0   4.244226e-13  MPI-ESM-1-2-HAM  
1   4.217326e-13  MPI-ESM-1-2-HAM  
2   4.498125e-13  MPI-ESM-1-2-HAM  
3   4.251282e-13  MPI-ESM-1-2-HAM  
4   4.270161e-13  MPI-ESM-1-2-HAM

In [4]:
# Checking data types of the columns
df.dtypes

time              object
lat_min          float64
lat_max          float64
lon_min          float64
lon_max          float64
rain (mm/day)    float64
model             object
dtype: object

In [5]:
#Checking if the max and min values of the float64-type columns are within the range of `float32`
df_numeric_summary = df.describe().loc[["min", "max"], :].T
df_numeric_summary["within float32 range"] = df_numeric_summary[["min", "max"]].apply(lambda x:
                                                                                      "True" if (x["min"] > np.finfo(np.float32).min
                                                                                                 and x["max"] < np.finfo(np.float32).max)
                                                                                             else "False", axis=1)
df_numeric_summary

min         max within float32 range
lat_min       -3.646739e+01  -29.900000                 True
lat_max       -3.600000e+01  -27.906064                 True
lon_min        1.406250e+02  153.750000                 True
lon_max        1.412500e+02  155.625000                 True
rain (mm/day) -3.807373e-12  432.939515                 True

In [6]:
print(f"Memory usage with float64 (default): {df[['lat_min','lat_max','lon_min', 'lon_max', 'rain (mm/day)']].memory_usage().sum() / 1e6:.2f} MB")
print(f"Memory usage with float32: {df[['lat_min','lat_max','lon_min', 'lon_max', 'rain (mm/day)']].astype('float32', errors='ignore').memory_usage().sum() / 1e6:.2f} MB")

Memory usage with float64 (default): 2498.71 MB
Memory usage with float32: 1249.36 MB


In [7]:
print(f"Memory usage with object (default) for the time column: {df[['time']].memory_usage().sum() / 1e6:.2f} MB")
print(f"Memory usage with datetime64 for the time column: {df[['time']].astype('datetime64[ns]', errors='ignore').memory_usage().sum() / 1e6:.2f} MB")

Memory usage with object (default) for the time column: 499.74 MB
Memory usage with datetime64 for the time column: 499.74 MB


In [8]:
# Checking the `model` column's unique values
df.model.unique()

array(['MPI-ESM-1-2-HAM', 'AWI-ESM-1-1-LR', 'NorESM2-LM', 'ACCESS-CM2',
       'FGOALS-f3-L', 'CMCC-CM2-HR4', 'MRI-ESM2-0', 'GFDL-CM4',
       'BCC-CSM2-MR', 'EC-Earth3-Veg-LR', 'CMCC-ESM2', 'NESM3',
       'MPI-ESM1-2-LR', 'ACCESS-ESM1-5', 'FGOALS-g3', 'INM-CM4-8',
       'MPI-ESM1-2-HR', 'TaiESM1', 'NorESM2-MM', 'CMCC-CM2-SR5',
       'KIOST-ESM', 'INM-CM5-0', 'MIROC6', 'BCC-ESM1', 'GFDL-ESM4',
       'CanESM5', 'SAM0-UNICON'], dtype=object)

In [9]:
print(f"Memory usage with object (default) for the model column: {df[['model']].memory_usage().sum() / 1e6:.2f} MB")
print(f"Memory usage with string for the model column: {df[['model']].astype('string', errors='ignore').memory_usage().sum() / 1e6:.2f} MB")

Memory usage with object (default) for the model column: 499.74 MB
Memory usage with string for the model column: 499.74 MB


In [10]:
# Calculating memory savings for each column
df_memory_savings = pd.DataFrame(columns = df.columns).T.reset_index().rename(columns={'index': 'column name'})

list_savings = []
savings = 0
for col in df.columns:
    if df[col].dtypes == "float64":
        savings = (df[[col]].memory_usage().sum() / 1e6) - (df[[col]].astype('float32', errors='ignore').memory_usage().sum() / 1e6)
        list_savings.append(savings)
        savings = 0
    elif col == "time":
        savings = (df[["time"]].memory_usage().sum() / 1e6) - (df[[col]].astype('datetime64[ns]', errors='ignore').memory_usage().sum() / 1e6)
        list_savings.append(savings)
        savings = 0
    else:
        savings = (df[["model"]].memory_usage().sum() / 1e6) - (df[[col]].astype('string', errors='ignore').memory_usage().sum() / 1e6)
        list_savings.append(savings)
        savings = 0

df_memory_savings['memory savings'] = pd.Series(list_savings).values
df_memory_savings

column name  memory savings
0           time        0.000000
1        lat_min      249.871372
2        lat_max      249.871372
3        lon_min      249.871372
4        lon_max      249.871372
5  rain (mm/day)      249.871372
6          model        0.000000

In [11]:
print("\033[1m" + f"Memory savings due to changing the data types of the columns: {df_memory_savings['memory savings'].sum():.2f} MB")

Memory savings due to changing the data types of the columns: 1249.36 MB


# Load just columns we want

In [12]:
print(f"Memory usage with all columns: {df[['time', 'lat_min','lat_max','lon_min', 'lon_max', 'rain (mm/day)', 'model']].memory_usage().sum() / 1e6:.2f} MB")

Memory usage with all columns: 3498.20 MB


In [13]:
# Dropping `lat_min` and `lon_min` columns
df_reduced = df.drop(['lat_min', 'lon_min'], axis=1);
print(f"Memory usage of the reduced dataframe: {df_reduced.memory_usage().sum() / 1e6:.2f} MB")
df_reduced.head()

Memory usage of the reduced dataframe: 2498.71 MB


time    lat_max   lon_max  rain (mm/day)            model
0  1889-01-01 12:00:00 -33.574619  143.4375   4.244226e-13  MPI-ESM-1-2-HAM
1  1889-01-02 12:00:00 -33.574619  143.4375   4.217326e-13  MPI-ESM-1-2-HAM
2  1889-01-03 12:00:00 -33.574619  143.4375   4.498125e-13  MPI-ESM-1-2-HAM
3  1889-01-04 12:00:00 -33.574619  143.4375   4.251282e-13  MPI-ESM-1-2-HAM
4  1889-01-05 12:00:00 -33.574619  143.4375   4.270161e-13  MPI-ESM-1-2-HAM

In [14]:
print("\033[1m" + f"Memory savings due to loading not all columns: {df[['lat_min', 'lon_min']].memory_usage().sum() / 1e6:.2f} MB")

Memory savings due to loading not all columns: 999.49 MB


## Conclusion

Our team decided to try two approaches to reduce memory usage when performing the EDA using Python:

**1. Change the data types of the columns in the dataframe.**
<br/> There are 2 columns of the `object` type (`time` and `model`) and 5 columns of the `float64` type (`lat_min`, `lat_max`, `lon_min`, `lon_max`, and `rain (mm/day)`).

The summary table for the numeric columns shows that all of them are whithin the range of the `float32` data type, i.e. the numeric columns' highest values do not exceed the `float32` maximum number, and their lowest values are not less than the `float32` minimum. Changing the data type for these columns into the `float32` type leads to the two-fold decrease in memory usage, i.e. the reduction from 2498.71 MB to 1249.36 MB. It is related to the fact that `float64` allocates twice as much memory as `float32`, since `float64` can store much larger numbers than `float32`.

Regarding the `time` and `model` columns, 499.74 MB of memory was allocated for each of the columns. Interestingly, changing the data type for the `time` and `model` columns to `datetime64[ns]` and `string` respectively did not change the amount of memory used.

Based on the above, we can conclude that the `datetime64`, `string`, and `object` data types allocate the same amount of memory for storing data. For this reason, in order to reduce memory usage it is reasonable to change the data type for numeric columns.

In this particular case, **memory savings** due to changing the data types of the numeric columns are 1249.36 MB.

**2. Load some specific columns we interested in.**

For the sake of the task, we suppose that now we are not interested in the `lat_min` and `lon_min` columns, i.e. these columns do not play a big role in our prediction problem. Dropping the specified columns reduced the memory usage from 3498.20 MB to 2498.71 MB.

This shows the efficiency of loading only those columns that are necessary for a specific task in terms of memory usage when working with big data.

Here, **memory savings** due to loading not all columns (by dropping two columns) are 999.49 MB.